src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php

```diff
@@ -57,7 +62,24 @@ public function onMessage(MessageEvent $event): void
         }
 
         if ($this->recipients) {
-            $event->getEnvelope()->setRecipients($this->recipients);
+            $recipients = $this->recipients;
+            if ($this->allowedRecipients) {
+                $recipients = [];
+                $addDefaults = false;
+                foreach ($event->getEnvelope()->getRecipients() as $recipient) {
+                    foreach ($this->allowedRecipients as $allowedRecipient) {
+                        if (preg_match($allowedRecipient, $recipient->getAddress())) {
+                            $recipients[] = $recipient;
+                            continue 2;
+                        }
+                    }
+                    $addDefaults = true;
+                }
+                if ($addDefaults) {
+                    $recipients = array_merge($recipients, $this->recipients);
+                }
+            }
```

Comment:
Looking at this logic: if one recipient isn't allowed, then we send to all $recipents, but if all recipients are allowed, then we don't send to $recipents. Shouldn't we always send to $recipents and in addition also send to allowed recipients? (and possibly dedup?)

Reply:
okay, done ✅